nni/algorithms/compression/pytorch/pruning/transformer_pruner.py

```diff
@@ -0,0 +1,321 @@
+            assert len(group) == 4, errmsg + ': each group must have four weights'
+            assert group[0].module.weight.size() == group[1].module.weight.size() and \
+                group[1].module.weight.size() == group[2].module.weight.size(), \
+                errmsg + ': the dimensions of Q, K, V projection matrices must be the same '
+            assert group[0].module.weight.size()[0] == group[3].module.weight.size()[1], \
+                errmsg + ': the dimension of attention results must match with input for output projection'
+            assert group[0].config['sparsity'] == group[1].config['sparsity'] == \
+                   group[2].config['sparsity'] == group[3].config['sparsity'], \
+                errmsg + ': the sparsity of matrices in the same layer must be the same'
+            if sparsity is None:
+                sparsity = group[0].config['sparsity']
+            if self.global_sort:
+                assert sparsity == group[0].config['sparsity'], \
+                    errmsg + ': for global_sort=True, the sparsity for all modules must be the same'
+            t = group[0].module.weight.size(0) / self.head_hidden_dim
+            assert t % 1 == 0, errmsg + ': head_hidden_dim must be a divisor of the output dimension of the ' \
+                                        'projection weights'
+
+    def remove_ungrouped_modules(self):
+        """
+        Remove non-attention weights that might be mistakenly captured by a simplified config_list.
+        Also update the corresponding list of layer information (self.modules_to_compress)
+        """
+        care_of_modules = set([x for layer in self.masking_groups for x in layer])
+
+        modules_wrapper_new, modules_to_compress_new = [], []
+        for wrapper, layer_info in zip(self.modules_wrapper, self.modules_to_compress):
+            if wrapper in care_of_modules:
+                modules_wrapper_new.append(wrapper)
+                modules_to_compress_new.append(layer_info)
+
+        self.modules_wrapper = modules_wrapper_new
+        self.modules_to_compress = modules_to_compress_new
+
+    def validate_config(self, model, config_list):
+        """
+        Parameters
+        ----------
+        model : torch.nn.Module
+            Model to be pruned
+        config_list : list
+            List on pruning configs
+        """
+        schema = CompressorSchema([{
+            'sparsity': And(float, lambda n: 0 < n < 1),
+            Optional('op_types'): [str],
+            Optional('op_names'): [str]
+        }], model, logger)
+
+        schema.validate(config_list)
+
+    def compress(self):
+        for pruning_iter in range(self.num_iterations):
+            if self.ranking_criterion in ['l1_activation', 'l2_activation', 'taylorfo']:
+                training = self.bound_model.training
+                self.bound_model.eval()
+                self._trainer(self.bound_model, optimizer=self._optimizer, criterion=self._criterion, epoch=0)
```

Comment:
why activation based criterion need training?

Reply:
We need to run a `forward` once to collect activations and (optionally) `backward` once to collect gradients (for `taylorfo`). This might be problematic if the trainer provided by the user also do parameter updates (i.e., `optimizer.step`). In the docs and example I mention that the user can write the trainer so that when `epoch=0` no parameter update is performed. The pruner also guarantees that when it needs to collect data, it always passes 0 to the parameter `epoch`. When it uses the trainer for finetuning, it never passes 0 to `epoch`.

Reply:
Or maybe we can pass -1 instead of 0.

Reply:
fixed; now the pruner passes `None` to the trainer's `epoch` argument when it is using this trainer to collect data.